### ISSUE TITLE
Soundness bug: Unsafe `init` function fails to properly specify invariants on memory region

### ISSUE THREAD

**Issue by TheButlah** [esp-rs/esp-alloc#6]
`EspHeap` currently uses `linked_list_allocator::Heap` in its implementation. The `init` function calls into the linked list's alloc function. However, it fails to properly document all of the invariants that that function requires. 

[here](https://docs.rs/linked_list_allocator/latest/linked_list_allocator/struct.Heap.html#safety) are the three safety invariants from linked list allocator:
* This function must be called at most once and must only be used on an empty heap.
* The bottom address must be valid and the memory in the [heap_bottom, heap_bottom + heap_size) range must not be used for anything else. This function is unsafe because it can cause undefined behavior if the given address is invalid.
* The provided memory range must be valid for the 'static lifetime.

Presumably due to the lack of this documentation of invariants, the `esp-alloc` heap is being called in an unsound way in several places:
* In [esp-template](https://github.com/esp-rs/esp-template/blob/968ce7469b04852eac4e3aef19cb9619ec7bf8a6/src/main.rs#L23) and therefore in anyone using the template
* In [esp-wifi](https://github.com/esp-rs/esp-wifi/blob/3f21a58eea26f0339a4dd967209d2cf41cf4f3c6/src/lib.rs#L84) (fixed by [#](https://github.com/esp-rs/esp-wifi/pull/73)

**Comment by MabezDev** [comment]
> The bottom address must be valid and the memory in the [heap_bottom, heap_bottom + heap_size) range must not be used for anything else. This function is unsafe because it can cause undefined behavior if the given address is invalid.

I believe this is only being violated on the esp32c3, because there is no heap_end symbol from the linker script. I think if we add this symbol we can guarantee we'll have an untouched block of memory to play with. Thoughts @bjoernQ, @jessebraham ?

**Comment by TheButlah** [comment]
why do this with a linker script (to many, linker scripts are magic) when instead it can be done with:
```rust
static mut HEAP_DATA: [MaybeUninit<u8>; HEAP_SIZE] = [MaybeUninit::new(0u8); HEAP_SIZE];
```
To me, reaching for a linker script seems unecessary when we can express the same idea in standard rust code.

**Comment by bjoernQ** [comment]
Probably we should just offer a second way to initialize from a mutable slice and also add some more documentation to the existing init function (while it's already marked unsafe)

The good thing about using the symbols provided by the linker is, we can just _use all the remaining memory for the heap_ while with an explicit slice we need to come up with some size and at some point the linker might start complaining that it doesn't fit anymore (maybe by just adding a dependency)

TL;DR lets add some more documentation to the current init function and provide a second one
